lib/WorkflowMain.groovy

```diff
@@ -2,92 +2,247 @@
+            }
+        }
+
+        def stages = Processes.getRunStages(
+            params.processes_include,
+            params.processes_exclude,
+            params.processes_manual,
+            log,
+        )
+
+        if (!params.containsKey('ref_data_virusbreakenddb_path') && stages.virusinterpreter && run_mode === Constants.RunMode.WGTS){
+            params.ref_data_virusbreakenddb_path = Constants.VIRUSBREAKENDDB_PATH
+        }
+
+        if (!params.containsKey('ref_data_hla_slice_bed') && stages.lilac) {
+            if (params.genome_version.toString() == '38' && params.genome_type == 'alt') {
+                params.ref_data_hla_slice_bed = Constants.HLA_SLICE_BED_GRCH38_ALT_PATH
+            }
         }
+
+        // Final point to set any default to avoid access to undefined parameters during nf-validation
+        if (!params.containsKey('panel')) { params.panel = null }
+        if (!params.containsKey('ref_data_genome_alt')) { params.ref_data_genome_alt = null }
+        if (!params.containsKey('ref_data_genome_gtf')) { params.ref_data_genome_gtf = null }
+        if (!params.containsKey('ref_data_hla_slice_bed')) { params.ref_data_hla_slice_bed = null }
+        if (!params.containsKey('ref_data_panel_data_path')) { params.ref_data_panel_data_path = null }
+        if (!params.containsKey('ref_data_virusbreakenddb_path')) { params.ref_data_virusbreakenddb_path = null }
+
     }
+
     //
-    // Get attribute from genome config file e.g. fasta
+    // Check and validate parameters
     //
-    public static Object getGenomeAttribute(params, attribute) {
-        if (params.genomes && params.genome && params.genomes.containsKey(params.genome)) {
-            if (params.genomes[ params.genome ].containsKey(attribute)) {
-                return params.genomes[ params.genome ][ attribute ]
+    public static void validateParams(params, log) {
+
+        // Common parameters
+
+        if (!params.genome) {
+            log.error "\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n" +
+                "  Genome must be set using the --genome CLI argument or in a configuration file.\n" +
+                "  Currently, the available genome are:\n" +
+                "  ${params.genomes.keySet().join(", ")}\n" +
+                "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+            Nextflow.exit(1)
+        } else if (!params.genomes.containsKey(params.genome)) {
+            log.error "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n" +
+                "  Genome '${params.genome}' not found in any config files provided to the pipeline.\n" +
+                "  Currently, the available genome are:\n" +
+                "  ${params.genomes.keySet().join(", ")}\n" +
+                "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+            Nextflow.exit(1)
+        }
```

Comment:
have you tried using the nextflow_schema.json to validate all these parameters instead? You can provide regexs and list etc there directly

Reply:
Since users are allowed to provide their own custom genome as long as they defined it in `params.genomes` via a configuration file, effectively any value can then be provided to the `--genome` argument and considered valid under the aforementioned condition. And as far as I understand the validation available from within nextflow_schema.json isn't useful in this scenario.